packages/vite/src/node/http.ts

```diff
@@ -183,3 +184,27 @@ export async function httpServerStart(
     })
   })
 }
+
+export function setClientErrorHandler(
+  server: HttpServer,
+  logger: Logger
+): void {
+  server.on('clientError', (err, socket) => {
+    if ((err as any).code === 'HPE_HEADER_OVERFLOW') {
+      if (!socket.writableEnded) {
+        socket.end(
+          'HTTP/1.1 431 Request Header Fields Too Large\r\n\r\n' +
+            'Request Header was too large. Node.js limits request header size. ' +
+            'Use https://nodejs.org/api/cli.html#--max-http-header-sizesize to change max header size.'
+        )
+      }
+      logger.warn(
+        colors.yellow(
+          'Server / WS server responded with "431 Request Header Fields Too Large." ' +
+            'Node.js limits request header size and the request was dropped. ' +
+            'Use https://nodejs.org/api/cli.html#--max-http-header-sizesize to change max header size.'
+        )
+      )
+    }
+  })
```

Comment:
I think it's a bit specific that we only handle this specific error, maybe we should log the `err` generically instead? (not sure if that provides any message).

Reply:
`clientError` event seems to be called with [400 and 431 (this one)](https://nodejs.org/api/http.html#event-clienterror) but I'm not sure if we should output a warning for 400 too.

The `err` looks like this if it is output directly.
![image](https://user-images.githubusercontent.com/49056869/180614062-20f8f74e-a5b2-4350-ad8a-2c174ff50972.png)

Maybe something like this is better?
```suggestion
  server.on('clientError', (err, socket) => {
    const code = (err as any).code
    if (code === 'HPE_HEADER_OVERFLOW') {
      logger.warn(
        colors.yellow(
          'Server / WS server responded with an error. ' +
            `${err.message} (code: ${code})`
        )
      )
    }
  })
```

Reply:
I think the custom error message here is useful. Maybe something in between this two?
Another option is now that you started a troubleshooting page, we could start adding these longer explanations there, linking to it from the error instead of needing these messages that with time could add up to the size in Vite.

Reply:
Thanks for the explanation @sapphi-red. I think the original PR makes sense then given two errors only happen here. It's strange that Node gives a cryptic error though, and my concern is with patak too that this increases the size for a very specific case.

I like the idea of putting this in the troubleshooting page too, we could probably elaborate more there too.

Reply:
Sounds nice. I've move the details to troubleshooting page 👍

Reply:
Maybe we could discuss a bit in today's meeting. We could start moving other messages to the new guide, and add a section in the CONTRIBUTING guide to guide other contributors to do the same from now on. Maybe these should be in a section that only list errors, like https://yarnpkg.com/advanced/error-codes

Reply:
FYI, how React team built their error code system: https://reactjs.org/blog/2016/07/11/introducing-reacts-error-code-system.html

Reply:
hi @sapphi-red ,in the above code, the socket is not closed, in my case, it will hang up the webpage untill timeout reached. Could you please help me review this pull request? #9816